query-engine/query-structure/src/query_arguments.rs

```diff
@@ -106,7 +106,30 @@ impl QueryArguments {
             .connector
             .has_capability(ConnectorCapability::DistinctOn);
 
-        has_distinct_feature && connector_can_distinct_in_db && self.order_by.is_empty()
+        has_distinct_feature && connector_can_distinct_in_db && self.can_orderby_with_in_db_distinct()
+    }
+
+    fn can_orderby_with_in_db_distinct(&self) -> bool {
+        match &self.distinct {
+            Some(distinct) => {
+                let dist_len = distinct.as_fields().len();
+
+                if self.order_by.len() <= dist_len {
+                    Self::ord_matches_distinct(&self.order_by, distinct)
+                } else {
+                    let initial_order = &self.order_by[0..dist_len];
+
+                    Self::ord_matches_distinct(initial_order, distinct)
+                }
+            }
+            None => unreachable!(),
+        }
+    }
+
+    fn ord_matches_distinct(order_by: &[OrderBy], distinct: &FieldSelection) -> bool {
+        order_by
+            .iter()
+            .all(|ord| distinct.contains(ord.field().unwrap().name()))
```

Comment:
This is going to panic if there's an order by that doesn't point to a scalar field. It should return `false` if that's the case.

Reply:
I can't figure out how this is a reachable state;
This is my understanding from a PSL perspective:
```rs
enum FieldType {
    Model(ast::ModelId),
    Scalar(ScalarFieldType),
}
```

If we examine the `Scalar` option, this includes the following:
```rs
pub enum ScalarFieldType {
    /// A composite type
    CompositeType(ast::CompositeTypeId),
    /// An enum
    Enum(ast::EnumId),
    /// A Prisma scalar type
    BuiltInScalar(ScalarType),
    /// An `Unsupported("...")` type
    Unsupported(UnsupportedType),
}
```

We can ignore `CompositeType` as this is not relevant for MongoDB; we can ignore `Unsupported` as fields with this type aren't included in the client generation.
Enum seems to be treated as a scalar given the above information and that the following query does _not_ panic.
```graphql
{
    findManyUser(orderBy: { status: desc }, distinct: [status])
    { id, first_name, status }
}"
```

With the following data:
```graphql
{ id: 1, first_name: "Joe", last_name: "Doe", email: "1" status: Alive },
{ id: 2, first_name: "Bro", last_name: "Doe", email: "2" status: Alive }
```
Yields:
```graphql
{ "data": { "findManyUser": [{ "id": 1, "first_name": "Joe", "status": "Alive" }] } }
```

And then finally, we have the following as Scalars.
```rs
pub enum ScalarType {
    Int,
    BigInt,
    Float,
    Boolean,
    String,
    DateTime,
    Json,
    Bytes,
    Decimal,
}
```

So this seems to just leave `Model`, however, if we look at what's available on generation when referencing Models as field types, i.e.
```prisma
model User {
  id         Int    @id
  first_name String
  post       Post?
}

model Post {
  id        Int    @id
  title     String
  author_id Int
  author    User   @relation(fields: [userId], references: [id])
  userId    Int    @unique
}
```
we can only access the following:
```ts
prisma.post.findMany({
  distinct: [...] // author_id | id | title | userId
})

const u = prisma.user.findMany({
  distinct: [...], // first_name | id
});
```

---

Just for my own understanding is this a possible state? either:
- within current implementation, or
- something that we don't support _yet_ but could in the future (and therefore something to future-proof for)

Reply:
`OrderBy::field` returns None for the following variants:

https://github.com/prisma/prisma-engines/blob/01aad9b63c8d574cc270d2b09461e920d19986e6/query-engine/query-structure/src/order_by.rs#L52-L59

None of your examples reaches this case. You can reach these code paths by:
- Adding a to-many relation to your datamodel and trying to sort by the to-many side of it
- Enabling the `fullTextSearch` preview feature and ordering by `_relevance`